ISPN-11393 statistics, metrics, jmx

oraNod opened the pull request:
https://issues.redhat.com/browse/ISPN-11393 

Needs backport to 10.1.x

anistor commented:
Looking...

anistor commented on documentation/src/main/asciidoc/topics/code_examples/ConfigStats.java:
```diff
@@ -0,0 +1,5 @@
+GlobalConfiguration globalConfig = new GlobalConfigurationBuilder()
+  .statistics().enable()
+  .metrics().gauges().histograms()
```
.metrics().gauges(true).histograms(true)

anistor commented on documentation/src/main/asciidoc/stories/assembly_server_statistics.adoc:
```diff
@@ -0,0 +1,17 @@
+[id='statistics']
+:context: server
+= Configuring Statistics, Metrics, and JMX
+Enable statistics that {brandname} exports to a metrics endpoint or via JMX
```
``` exports to a metrics endpoint```  --> ``` exports to a MicroProfile Metrics endpoint```

anistor commented on documentation/src/main/asciidoc/topics/code_examples/HotRodClientInstance.java:
```diff
@@ -3,7 +3,6 @@
 cb.marshaller(new org.infinispan.commons.marshall.ProtoStreamMarshaller())
   .statistics()
       .enable()
-      .jmxDomain("org.infinispan")
```
jmxDomain is not deprecated in HR client. This line can be kept, but maybe I would use something different than "org.infinispan", which is the default anyway, so it's not useful.

anistor commented on core/src/main/resources/schema/infinispan-config-11.0.xsd:
```diff
@@ -270,12 +270,16 @@
         <xs:annotation>
-          <xs:documentation>Defines JMX management details.</xs:documentation>
+          <xs:documentation>
+            Enables and configures JMX monitoring.
```
monitoring and management

anistor commented on documentation/src/main/asciidoc/topics/proc_enabling_metrics.adoc:
```diff
@@ -0,0 +1,44 @@
+  .build();
+----
+
+<1> exports statistics about the Cache Manager.
```
```exports statistics about ...``` --> ```computes statistics about ...```

anistor commented on documentation/src/main/asciidoc/topics/proc_enabling_metrics.adoc:
```diff
@@ -0,0 +1,44 @@
+----
+
+<1> exports statistics about the Cache Manager.
+<2> generates gauge and histogram metrics.
```
```generates gauge and histogram metrics```  --> ```exports the collected statistics as gauge and histogram metrics```

anistor commented on documentation/src/main/asciidoc/topics/proc_enabling_metrics.adoc:
```diff
@@ -0,0 +1,44 @@
+----
+GlobalConfiguration globalConfig = new GlobalConfigurationBuilder()
+  .statistics().enable() <1>
+  .metrics().gauges().histograms() <2>
```
.metrics().gauges(true).histograms(true)

anistor commented on documentation/src/main/asciidoc/topics/proc_enabling_statistics.adoc:
```diff
@@ -0,0 +1,48 @@
+  .build();
+----
+
+<1> exports statistics about the Cache Manager.
```
again, 'collects' instead of 'exports', here and on next line

anistor commented on documentation/src/main/asciidoc/topics/proc_enabling_statistics.adoc:
```diff
@@ -0,0 +1,48 @@
+[NOTE]
+====
+Enabling statistics for Cache Managers does not globally enable statistics for
+caches. This configuration only exports statistics for the Cache Manager.
```
collects statistics

anistor commented on documentation/src/main/asciidoc/topics/proc_collecting_metrics.adoc:
```diff
@@ -1,45 +1,37 @@
+
+.Prerequisites
+
+* Enable statistics. If you do not enable statistics, {brandname} provides `0` values for all metrics.
```
0 or in some cases -1

anistor commented on documentation/src/main/asciidoc/topics/proc_enabling_metrics.adoc:
```diff
@@ -0,0 +1,44 @@
+{brandname} is compatible with the Eclipse MicroProfile Metrics API and can
+generate gauge and histogram metrics.
+
+* {brandname} metrics are provided at the `/base` and `/vendor` scopes and are
```
{brandname} metrics are provided at the 'vendor' scope. Also, JVM-related metrics are provided in the 'base' scope.

anistor commented on core/src/main/resources/schema/infinispan-config-11.0.xsd:
```diff
@@ -641,21 +645,21 @@
       <xs:annotation>
         <xs:documentation>
-          Enables exporting of gauge metrics. These are exported by default if statistics are enabled.
+          Exports gauge metrics. If you enable statistics, gauges are exported by default.
```
I would rephrase ```Exports gauge metrics. Gauges are enabled by default. Statistics must also be enabled or else no metrics will actually be exported. ```

anistor commented on core/src/main/resources/schema/infinispan-config-11.0.xsd:
```diff
@@ -641,21 +645,21 @@
       <xs:annotation>
         <xs:documentation>
-          Enables exporting of histograms. These are costly to compute so are not enabled by default.
+          Exports histograms. Disabled by default because histograms require additional computation.
```
I would rephrase ```Exports histogram metrics. Histograms are not enabled by default because they require additional computation. Statistics must also be enabled or else no metrics will actually be exported. ```

anistor commented on core/src/main/resources/schema/infinispan-config-11.0.xsd:
```diff
@@ -672,28 +676,30 @@
       <xs:annotation>
         <xs:documentation>
-          If JMX statistics are enabled then all 'published' JMX objects will appear under this name. This is optional, if not specified an object name will be created for you by default.
+          If JMX statistics are enabled then all 'published' JMX objects  appear under this name. Optional, if not specified a default object name is created.
```
strange space between ```objects``` and ```appear```

anistor commented on documentation/src/main/asciidoc/topics/proc_configuring_hotrod_jmx.adoc:
```diff
@@ -13,8 +10,19 @@ Rod client statistics.
+clientBuilder
+   .statistics()
+    .enable() <1>
+   .jmx()
```
replace ```jmx().enable()``` with ```jmxEnable()```

anistor commented on documentation/src/main/asciidoc/topics/proc_configuring_hotrod_jmx.adoc:
```diff
@@ -13,8 +10,19 @@ Rod client statistics.
 [source,java]
 ----
-include::code_examples/HotRodJmx.java[]
+ConfigurationBuilder clientBuilder = new ConfigurationBuilder();
```
An ```import org.infinispan.client.hotrod.configuration.ConfigurationBuilder;``` should make it clear we're not referring to the ConfigurationBuilder from core

anistor commented on documentation/src/main/asciidoc/topics/proc_naming_cache_managers.adoc:
```diff
@@ -14,15 +14,21 @@ name, which results in a JMX MBean named `org.infinispan:type=CacheManager,name=
 ----
-include::code_examples/JmxStatisticsCacheManager.java[]
+GlobalConfiguration globalConfig = new GlobalConfigurationBuilder()
+  .jmx().enable()
```
Lines 30 should be before 29 or it won't compile.

oraNod commented:
> I've added another bunch of comments. I think I'm about 90% done. Please fix these and then I'll read everything again on the rendered doc.

thanks @anistor this is a really good review. I've learned a few things here. the changes are in and I've updated the rendered doc.

anistor commented:
Merged in master @oraNod ! Thanks!

anistor commented:
While reviewing the docs (and also looking at what the code actually does, just to make sure) I found this https://issues.redhat.com/browse/ISPN-11463
